odoo/tools/misc.py

```diff
@@ -1641,3 +1641,14 @@ def hmac(env, scope, message, hash_function=hashlib.sha256):
         message.encode(),
         hash_function,
     ).hexdigest()
+
+
+bots = ["bot", "crawl", "slurp", "spider", "curl", "wget", "facebookexternalhit", "whatsapp", "trendsmapresolver", "pinterest", "instagram"]
+def is_a_bot(request):
+    # We don't use regexp and ustr voluntarily
+    # timeit has been done to check the optimum method
```

Comment:
not very relevant comment

Reply:
Isn't that comment relevant? It will prevent further change to revert what was done as one would be tempted to use a regex here

Reply:
I do not think, it's like each time you do `"str" in a` you said "we didn't use regex, or a library, or a C extension to do that because it's slower" (and it's obvious that regex will be slower...)

Reply:
I did Edit my message the same time you wrote yours :yum: 

It is indeed obvious, and most of the time it does not matter that much, that's why some colleagues will always prefer to change that kind of code to a regex when doing a quick fix or improvement on such code, which is fine most of the time.

But in this particular case, this method is called on every http request during the dispatching, where any gain is obviously very important.
That's why this comment seems relevant here :-)

Just my opinion tho.

Reply:
> I do not think, it's like each time you do "str" in a you said "we didn't use regex, or a library, or a C extension to do that because it's slower" (and it's obvious that regex will be slower...)

It's not necessarily obvious that an alternate regex will be slower than an O(n^2) containment check though.

Reply:
```
any(bot in user_agent for bot in cls.bots)
```
is O(n) not O(n^2)

Reply:
O(m*n) technically, it's the number of bots * the length of the UAS (UAS tend to be quite long).

Reply:
with this move, you cannot properly override the function :fearful:

Reply:
@JKE-be noted the same issue in https://github.com/odoo/odoo/pull/78806/files/7372363bbc817bcc73e18622641e37d7f90fd4b4#diff-61162ac65633a1c7b054fc83ce1813f1a7984e3169ff36021713ef441f62a208